TempoWorld/Source/TempoDebris/Private/PCGAssignRandomMeshAttribute.cpp

```diff
@@ -0,0 +1,249 @@
+	InputPinProperty.SetRequiredPin();
+
+	return PinProperties;
+}
+
+TArray<FPCGPinProperties> UPCGAssignRandomMeshAttributeSettings::OutputPinProperties() const
+{
+	TArray<FPCGPinProperties> PinProperties;
+	PinProperties.Emplace(PCGPinConstants::DefaultOutputLabel, EPCGDataType::Any);
+
+	return PinProperties;
+}
+
+FPCGElementPtr UPCGAssignRandomMeshAttributeSettings::CreateElement() const
+{
+	return MakeShared<FPCGAssignRandomMeshAttribute>();
+}
+
+EPCGDataType UPCGAssignRandomMeshAttributeSettings::GetCurrentPinTypes(const UPCGPin* InPin) const
+{
+	check(InPin);
+	if (!InPin->IsOutputPin())
+	{
+		return Super::GetCurrentPinTypes(InPin);
+	}
+
+	// Output pin narrows to union of inputs on first pin
+	const EPCGDataType PrimaryInputType = GetTypeUnionOfIncidentEdges(PCGPinConstants::DefaultInputLabel);
+	return (PrimaryInputType != EPCGDataType::None) ? PrimaryInputType : EPCGDataType::Param; // No input (None) means param.
+}
+
+FString UPCGAssignRandomMeshAttributeSettings::GetAdditionalTitleInformation() const
+{
+	return FString::Printf(TEXT("%s[ %s ]"), DataTable ? *DataTable->GetFName().ToString() : TEXT("None"), *RowName.ToString());
+}
+
+bool FPCGAssignRandomMeshAttribute::PrepareDataInternal(FPCGContext* Context) const
+{
+	TRACE_CPUPROFILER_EVENT_SCOPE(FPCGAssignRandomMeshAttribute::PrepareDataInternal);
+
+	check(Context);
+
+	const UPCGAssignRandomMeshAttributeSettings* Settings = Context->GetInputSettings<UPCGAssignRandomMeshAttributeSettings>();
+	check(Settings);
+
+	if (Settings->DataTable.IsNull())
+	{
+		return true;
+	}
+
+	FPCGAssignRandomMeshAttributeContext* ThisContext = static_cast<FPCGAssignRandomMeshAttributeContext*>(Context);
+
+	if (!ThisContext->WasLoadRequested())
+	{
+		return ThisContext->RequestResourceLoad(ThisContext, { Settings->DataTable.ToSoftObjectPath() }, !Settings->bSynchronousLoad);
+	}
+
+	return true;
+}
+
+bool FPCGAssignRandomMeshAttribute::ExecuteInternal(FPCGContext* Context) const
+{
+	TRACE_CPUPROFILER_EVENT_SCOPE(FPCGAssignRandomMeshAttribute::Execute);
+
+	check(Context);
+	const UPCGAssignRandomMeshAttributeSettings* Settings = Context->GetInputSettings<UPCGAssignRandomMeshAttributeSettings>();
+
+	check(Settings);
+	const TSoftObjectPtr<UDataTable> DataTablePtr = Settings->DataTable;
+	const FName RowName = Settings->RowName;
+
+	const UDataTable* DataTable = DataTablePtr.Get();
+	if (!DataTable)
+	{
+		PCGE_LOG(Warning, GraphAndLog, LOCTEXT("DataTableInvalid", "FPCGAssignRandomMeshAttribute: Data table is invalid"));
+		return true;
+	}
+
+	FStaticMeshDistribution* Row = DataTable->FindRow<FStaticMeshDistribution>(RowName, TEXT(""), false);
+	if (!Row)
+	{
+		PCGE_LOG(Error, GraphAndLog, FText::Format(LOCTEXT("NoRowFound", "FPCGAssignRandomMeshAttribute: Data table '{0}' has no row named '{1}'"), FText::FromString(DataTable->GetPathName()), FText::FromName(RowName)));
+		return true;
+	}
+
+	TMap<TSoftObjectPtr<UStaticMesh>, float> MeshDistribution;
```

Comment:
Sort the map

Reply:
I actually tested it and I *haven't* found any evidence that TMap iteration isn't deterministic:
https://github.com/tempo-sim/Tempo/tree/map_determinism_test

Still, I'm leery. So I'm sorting it.